### PR TITLE
fix: eliminate DNS rebinding TOCTOU vulnerability in SSRF-protected fetch

### LINKED ISSUE #80: feat: Mitigate DNS rebinding (TOCTOU) in SSRF-protected RSS fetch
## Context

**User Story:** As a platform operator, I want the RSS feed fetch to be immune to DNS rebinding attacks so that server-side request forgery (SSRF) cannot bypass our URL validation.

PR #77 introduced SSRF protection by validating URLs with `isSafeUrl()` (DNS resolution + private IP check) before fetching. However, there is a TOCTOU (Time-of-Check-Time-of-Use) gap: `isSafeUrl()` resolves DNS and validates the IPs, then `parsePodcastFeed()` calls `rss-parser`'s `parseURL()` which performs a **second, independent DNS resolution** via Node's `http.get`/`https.get`. An attacker controlling a DNS server can return a public IP for the first lookup and `127.0.0.1` for the second (DNS rebinding), bypassing all validation.

## Proposed Solution

Eliminate the TOCTOU gap by creating a custom `http.Agent` / `https.Agent` that:

1. Overrides `createConnection` to resolve DNS, check every resolved IP against `isPrivateIP()`, and only connect if all IPs are safe
2. Pins the validated IP for the actual TCP connection (no second DNS lookup occurs)
3. Is passed to `rss-parser` via its `requestOptions.agent` option

This collapses the validate-then-fetch into a single atomic operation at the socket level.

## Acceptance Criteria

- [ ] VERIFY: Given a URL whose DNS initially resolves to a public IP, when DNS is rebound to `127.0.0.1` between validation and fetch, then the connection is rejected before any data is sent
- [ ] VERIFY: Given a URL resolving to a public IP, when `parsePodcastFeed` fetches it, then the feed is parsed successfully (no regression)
- [ ] VERIFY: Given a URL resolving to multiple IPs (one public, one private), when the agent resolves DNS, then the connection is rejected
- [ ] VERIFY: `isSafeUrl()` is still called as the first-pass check (fail-fast before allocating agent resources)
- [ ] VERIFY: The custom agent respects the existing timeout configuration (60s)
- [ ] VERIFY: Unit tests cover the DNS-pinning agent behavior
- [ ] VERIFY: All existing tests pass without modification

## Edge Cases

- **Multiple A/AAAA records:** If any resolved IP is private, reject all -- do not try the next IP
- **IPv6-only hosts:** The agent must handle IPv6 addresses and apply the same `isPrivateIP()` checks
- **DNS resolution failure in agent:** Reject the connection (fail-closed), same as current `isSafeUrl` behavior
- **Connection timeout:** The agent should not extend the existing 60-second rss-parser timeout
- **Redirect bypass already handled:** `maxRedirects: 0` was set in PR #77, so the agent does not need to re-validate redirect targets (there are none)

### Files to Modify
- `src/lib/security.ts` -- add `createSsrfSafeAgent()` factory function that returns `{ httpAgent, httpsAgent }`
- `src/lib/rss.ts` -- import the agents from security.ts, pass them to the rss-parser constructor via `requestOptions`
- `src/lib/__tests__/security.test.ts` -- add tests for the DNS-pinning agent

### Approach

1. **Create `createSsrfSafeAgent()` in `src/lib/security.ts`:**
   - Extend `http.Agent` and `https.Agent`
   - Override `createConnection(options, callback)` to:
     a. Resolve the hostname with `dns.lookup(hostname, { all: true })`
     b. Check every resolved address with `isPrivateIP()`
     c. If any is private, destroy the socket and throw an error
     d. If all safe, set `options.host` to the first safe IP and call `super.createConnection()`
   - Export a factory: `createSsrfSafeAgent()` returning `{ httpAgent: http.Agent, httpsAgent: https.Agent }`

2. **Wire into rss-parser in `src/lib/rss.ts`:**
   - Import `createSsrfSafeAgent` from `@/lib/security`
   - Create agents at module scope (they are reusable across requests)
   - Since rss-parser selects `http.get` or `https.get` based on URL protocol and passes `requestOptions` directly, and Node's `http.get` uses the `agent` property from the options object, pass the appropriate agent. However, since feeds can be http or https, consider the alternative approach below.

3. **Alternative approach (recommended) -- fetch-then-parse wrapper:**
   - Instead of relying on rss-parser's `requestOptions`, fetch the XML yourself using Node's `http`/`https` with the safe agent, then call `parser.parseString(xml)` instead of `parser.parseURL(url)`. This gives full control over the HTTP request and avoids depending on rss-parser's internal request plumbing.
   - `parsePodcastFeed` fetches with the safe agent, gets the XML string, then `parser.parseString(xml)` parses it.
   - This is cleaner because rss-parser's `parseURL` method (in `node_modules/rss-parser/lib/parser.js:71-110`) merges `requestOptions` into `http.get` options, but selecting the right agent per-protocol at the constructor level is fragile.

4. **Add unit tests in `src/lib/__tests__/security.test.ts`:**
   - Test that the agent rejects connections to private IPs
   - Test that the agent allows connections to public IPs
   - Test that multiple resolved IPs with any private IP causes rejection
   - Test DNS failure causes rejection

### Patterns to Follow
- Follow the existing pattern in `src/lib/security.ts` where `isPrivateIP()` is the core check
- Follow the existing `promisify(dns.lookup)` pattern already used in `isSafeUrl()`
- rss-parser constructor options pattern is already established in `src/lib/rss.ts`

### Constraints
- **Must not break existing feeds:** All currently-working RSS feed URLs must continue to parse successfully
- **Must not increase latency significantly:** DNS resolution already happens in the current flow; the agent just combines it with the connection step
- **Must be compatible with rss-parser v3.x:** The library uses Node's built-in `http`/`https` modules, which support custom agents
- **No new dependencies:** Use only Node.js built-in `http`, `https`, `net`, and `dns` modules

### In Scope
- Custom DNS-pinning HTTP/HTTPS agent
- Integration with rss-parser's request pipeline
- Unit tests for the agent

### Out of Scope
- Replacing rss-parser with a different library
- Rate limiting DNS resolution
- DNS caching (Node handles this at the OS level)
- Protecting other HTTP calls in the app (this is RSS-specific)

## Dependencies

- Related to: #77 (Fix Blind SSRF in Podcast Subscription)
- Blocked by: nothing (PR #77 is the prerequisite and is already in-progress)
- Blocks: nothing

## Testing Requirements

- [ ] Unit tests for `createSsrfSafeAgent` or equivalent: private IP rejection, public IP pass-through, multi-record handling, DNS failure
- [ ] Existing `parsePodcastFeed` tests continue to pass
- [ ] Existing `isSafeUrl` tests continue to pass
- [ ] Integration-level test: mock DNS to return private IP during agent's `createConnection` and verify the fetch fails

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Add DNS-pinning undici Agent that eliminates the TOCTOU (time-of-check-to-time-of-use) vulnerability in `safeFetch()` where an attacker could DNS-rebind between `isSafeUrl()` validation and `fetch()`'s independent DNS resolution
- Custom `connect.lookup` resolves DNS once, validates ALL IPs against `isPrivateIP()`, and pins the validated IP for the TCP connection — no rebinding window
- Defense-in-depth preserved: `isSafeUrl()` pre-check retained for fast rejection; DNS-pinning agent provides connection-level guarantee
- Fail-closed: DNS errors, empty results, and any private IP in the resolved set all reject the connection
- TLS integrity maintained: `connect.lookup` only affects IP for TCP; hostname preserved for SNI/cert validation
- ADR-005 documents the undici `connect.lookup` approach and alternatives considered

## Changes

| File | Description |
|------|-------------|
| `src/lib/dns-pinning-agent.ts` | New module: `pinnedLookup`, `createDnsPinningAgent()`, `dnsPinningAgent` singleton |
| `src/lib/security.ts` | Import agent, add `dispatcher: dnsPinningAgent` to `fetch()` in safeFetch |
| `src/lib/__tests__/dns-pinning-agent.test.ts` | 12 unit tests: public/private IPv4/IPv6, mixed IPs, DNS failures, empty results, option preservation |
| `src/lib/__tests__/security.test.ts` | 4 integration tests: dispatcher passed, agent type, unsafe URL rejection, response body |
| `docs/adr/005-dns-pinning-ssrf-agent.md` | ADR documenting decision and alternatives |
| `CHANGELOG.md` | Entry under `[Unreleased] > Fixed` |

## Test plan

- [x] All 405 tests pass across 45 files
- [x] 12 new DNS-pinning agent unit tests cover: public IP allowed, `{ all: true }` flag, private IPv4/IPv6 rejected, link-local rejected, mixed IPs rejected, DNS failure fail-closed, empty results rejected, IPv4-mapped IPv6 rejected, caller options preserved, agent creation
- [x] 4 new safeFetch integration tests cover: dispatcher option passed, dnsPinningAgent is the dispatcher, unsafe URLs still rejected pre-fetch, response body returned
- [x] Lint clean, build successful
- [ ] Verify no regression on RSS feed subscription flow

Closes #80

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed a DNS rebinding (TOCTOU) vulnerability in SSRF protection by pinning DNS resolution to connections, preventing private-IP bypasses.

* **Behavioral Change**
  * Network fetches now enforce DNS-pinning for outbound requests while preserving existing URL safety checks and redirect behavior.

* **Tests**
  * Added unit and integration tests covering DNS resolution, rebinding scenarios, IP validation, and fetch integration.

* **Documentation**
  * Added design notes describing the DNS-pinning approach and fail-closed semantics.

### CODE REVIEW

File: docs/adr/005-dns-pinning-ssrf-agent.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   - Pass `{ dispatcher: dnsPinningAgent }` to each `fetch()` call.
+   - Keep the existing `isSafeUrl()` pre-check as defense-in-depth (fast rejection of obviously-bad URLs without establishing a TCP connection).
+
+3. **Tests: `src/lib/__tests__/dns-pinning-agent.test.ts`**
+   - Unit tests for the agent's DNS resolution and IP validation behavior.
+   - Tests for DNS rebinding scenarios (mock DNS returning private IP).
+   - Tests confirming TLS hostname verification is preserved.
+
+4. **Integration tests: update `src/lib/__tests__/rss-security.test.ts`**
+   - Add test cases verifying end-to-end protection against DNS rebinding through the `parsePodcastFeed` -> `safeFetch` -> DNS-pinning agent chain.
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply docs/adr/005-dns-pinning-ssrf-agent.md]: Fixed: Updated ADR section 4 to reference `src/lib/__tests__/security.test.ts` (the actual test location) instead of `rss-security.test.ts`.
